crates/holochain/src/test_utils/shared_values/mod.rs

```diff
@@ -0,0 +1,327 @@
+                                        .lock()
+                                        .await
+                                        .entry(key.to_string())
+                                        .or_default()
+                                        .clone()
+                                };
+
+                    notifier.notified().await;
+                    ignore_existing = false;
+
+                    localv1.num_waiters.fetch_sub(1, Ordering::SeqCst);
+                }
+            }
+            SharedValues::RemoteV1Client(_) => unimplemented!(),
+        }
+    }
+
+    /// Gets all values that have a matching key prefix; waits for `min_results` to become available if specified.
+    pub async fn get_pattern<T: for<'a> Deserialize<'a>>(
+        &mut self,
+        pattern: &str,
+        mut ignore_existing: bool,
+        maybe_min_results: Option<usize>,
+    ) -> Fallible<HashMap<String, T>> {
+        match self {
+            SharedValues::LocalV1(localv1) => {
+                loop {
+                    let notifier =
+                                // new scope so data_guard gets dropped before waiting for a notification
+                                {
+                                    let data_guard = localv1.data.lock().await;
+
+
+                                    if !ignore_existing {
+                                        let mut results: HashMap<String, T>  = Default::default();
+
+                                        for (key, value) in data_guard.iter() {
+                                            if key.matches(pattern).count() > 0 {
+                                                results.insert(key.to_string(), serde_json::from_str(&value)?);
+                                            }
+                                        }
+
+                                        if let Some(min_results) = maybe_min_results {
+                                            if results.len() >= min_results {
+                                                return Ok(results);
+                                            }
+                                        }
+                                    }
+
+                                    // get the notifier while still holding the data lock.
+                                    // this prevents a race between getting the notifier and a writer just writing something and sending notifications for it
+                                    localv1.num_waiters.fetch_add(1, Ordering::SeqCst);
+                                    localv1
+                                        .notification
+                                        .lock()
+                                        .await
+                                        .entry(pattern.to_string())
+                                        .or_default()
+                                        .clone()
+                                };
+
+                    notifier.notified().await;
```

Comment:
@holochain/core-dev i'd appreciate feedback on this potential race condition between the dropping of the guard and waiting for a notification:

1. task A drops the data guard, gets suspended by the scheduler
2. task B (in `fn put`) gets the data guard, calls the `notify_waiters` on all notifications _before_ task A gets into the `notifier.notified()` call
3. task B gets resumed and has already missed the notification

it'll be difficult to prove the above but it seems possible. ideally we'd have an async version of [std](https://doc.rust-lang.org/std/index.html)::[sync](https://doc.rust-lang.org/std/sync/index.html)::[Condvar](https://doc.rust-lang.org/std/sync/struct.Condvar.html#) but there's at least none in tokio.

Reply:
Without having looked at the code, I can say that it's certainly possible to call `notified()` after a notification has been sent. If the test is depending on that you could use a tokio barrier to ensur the task waits.